File: three-heart-land-run/core/src/gparap/games/threeheartlandrun/HUD.java
/*
 * Copyright 2024 gparap
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package gparap.games.threeheartlandrun;

import static com.badlogic.gdx.utils.Align.left;
import static com.badlogic.gdx.utils.Align.right;
import static java.lang.String.format;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.g2d.BitmapFont;
import com.badlogic.gdx.scenes.scene2d.Stage;
import com.badlogic.gdx.scenes.scene2d.ui.Label;
import com.badlogic.gdx.scenes.scene2d.ui.Table;
import com.badlogic.gdx.utils.Disposable;
import com.badlogic.gdx.utils.viewport.FillViewport;
import com.badlogic.gdx.utils.viewport.Viewport;

import java.util.Locale;

public class HUD implements Disposable {
    private final Stage stage;
    private int distance, score, highScore;
    private Label labelHighScoreL, labelHighScoreR, labelScoreL, labelScoreR, labelDistanceL, labelDistanceR;
    private final BitmapFont font;

    public HUD(ThreeHeartLandRun game) {
        Viewport viewport = new FillViewport(ThreeHeartLandRun.MIN_WIDTH, ThreeHeartLandRun.MIN_HEIGHT, game.getCamera());
        stage = new Stage(viewport, game.getSpriteBatch());
        font = new BitmapFont(Gdx.files.internal("fonts/test-font.fnt"));
        font.getData().setScale(0.5f);
        init();
        createLabels();
        stage.addActor(createTable());
    }

    public void setDistance(int distance) {
        this.distance = distance;
    }

    public void setScore(int score) {
        this.score = score;
    }

    public void setHighScore(int highScore) {
        this.highScore = highScore;
    }

    /**
     * Initializes all heads up display's variables.
     */
    private void init() {
        score = 0;
        distance = 0;
        highScore = 0;    //TODO: handle with preferences
    }

    /**
     * Creates all the labels that will be shown in the heads up display.
     */
    private void createLabels() {
        labelHighScoreL = new Label(format(Locale.getDefault(), "%s", "HIGH SCORE: "),
                new Label.LabelStyle(font, Color.WHITE));
        labelHighScoreR = new Label(format(Locale.getDefault(), "%3d", 0),
                new Label.LabelStyle(font, Color.WHITE));
        labelScoreL = new Label(format(Locale.getDefault(), "%s", "SCORE: "),
                new Label.LabelStyle(font, Color.WHITE));
        labelScoreR = new Label(format(Locale.getDefault(), "%3d", 0),
                new Label.LabelStyle(font, Color.WHITE));
        labelDistanceL = new Label(format(Locale.getDefault(), "%s", "DISTANCE: "),
                new Label.LabelStyle(font, Color.WHITE));
        labelDistanceR = new Label(format(Locale.getDefault(), "%3d%s", 0, " m"),
                new Label.LabelStyle(font, Color.WHITE));
    }

    /**
     * Create the table that will hold all labels that will be shown in the heads up display.
     *
     * @return table
     */
    private Table createTable() {
        Table table = new Table();
        float padLeft = 0.0f;
        float padTop = 0.0f;

        //adjust the table padding for different sizes
        switch(Gdx.app.getType()) {
            case Android:
                padLeft = 8;
                padTop = 8;
                if (Gdx.graphics.getDensity() < 2.5) {
                    padLeft = 32;
                    padTop = 12;
                }
                break;
            case Desktop:
                padLeft = 16;
                padTop = 16;
            case WebGL:
                //TODO("Not implemented yet.")
        }

        table.left().top().padLeft(padLeft).padTop(padTop);
        table.setFillParent(true);
        table.add(labelHighScoreL).align(right);
        table.add(labelHighScoreR).align(left);
        table.row();
        table.add(labelScoreL).align(right);
        table.add(labelScoreR).align(left);
        table.row();
        table.add(labelDistanceL).align(right);
        table.add(labelDistanceR).align(left);
        return table;
    }

    public void update() {
        //update display
        labelScoreR.setText(format(Locale.getDefault(), "%3d", score));
        labelDistanceR.setText(format(Locale.getDefault(), "%3d m", distance));
        labelHighScoreR.setText(format(Locale.getDefault(), "%3d", highScore));
    }

    public void render() {
        stage.draw();
    }

    /**
     * Releases all resources of the heads up display.
     */
    @Override
    public void dispose() {
        stage.dispose();
        font.dispose();
    }
}

File: falling/core/src/gparap/games/falling/HUD.kt
/*******************************
 * Katocheánian Gaming Studios *
 * Little Jerry's Friends      *
 * created by gparap           *
 *******************************/
package gparap.games.falling

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.graphics.g2d.BitmapFont
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.graphics.g2d.freetype.FreeTypeFontGenerator
import com.badlogic.gdx.graphics.g2d.freetype.FreeTypeFontGenerator.FreeTypeFontParameter
import com.badlogic.gdx.scenes.scene2d.Stage
import com.badlogic.gdx.scenes.scene2d.ui.Label
import com.badlogic.gdx.scenes.scene2d.ui.Label.LabelStyle
import com.badlogic.gdx.scenes.scene2d.ui.Table
import com.badlogic.gdx.utils.Align
import com.badlogic.gdx.utils.Disposable
import com.badlogic.gdx.utils.viewport.FitViewport
import java.util.*

/**
 * Heads-up display (presents game data to the user)
 */
class HUD(spriteBatch: SpriteBatch?) : Disposable {
    private val stage: Stage
    private var font: BitmapFont
    private var score = 0
    private var life = 0
    private var labelScoreL: Label? = null  //displays the label "SCORE:
    private var labelScoreR: Label? = null  //displays the actual score
    private var labelLifeL: Label? = null   //displays the label "LIFE: "
    private var labelLifeR: Label? = null   //displays the actual life

    fun setScore(score: Int) {
        this.score += score
    }

    fun setLife(life: Int) {
        this.life = life
    }

    init {
        val viewport = FitViewport(Gdx.graphics.width.toFloat(), Gdx.graphics.height.toFloat(), OrthographicCamera())
        stage = Stage(viewport, spriteBatch)
        font = BitmapFont(Gdx.files.internal(GameConstants.DEFAULT_FONT))
        score = 0
        life = 5
        createFont()
        createLabels()
        stage.addActor(createLabelsTable())
    }

    fun update() {
        labelScoreR?.setText(score)
        labelLifeR?.setText(life)
    }

    fun draw() {
        stage.draw()
    }

    /* Creates the font that will be shown in the heads-up display */
    private fun createFont() {
        val generator = FreeTypeFontGenerator(Gdx.files.internal(GameConstants.OPEN_TYPE_FONT))
        val parameter = FreeTypeFontParameter()
        parameter.size = getFontSize()
        parameter.color = Color.GOLD
        parameter.shadowColor = Color.BLACK
        parameter.shadowOffsetX = 1
        parameter.shadowOffsetY = 1
        font = generator.generateFont(parameter)
        generator.dispose()
    }

    /* Creates all the labels that will be shown in the heads-up display */
    private fun createLabels() {
        labelScoreL = Label(
            String.format(Locale.getDefault(), GameConstants.HUD_LABEL_FORMAT_STRING, GameConstants.HUD_LABEL_SCORE),
            LabelStyle(font, Color.GOLD)
        )
        labelScoreR = Label(
            String.format(Locale.getDefault(), GameConstants.HUD_LABEL_FORMAT_INTEGER, score),
            LabelStyle(font, Color.GOLD)
        )
        labelLifeL = Label(
            String.format(Locale.getDefault(), GameConstants.HUD_LABEL_FORMAT_STRING, GameConstants.HUD_LABEL_LIFE),
            LabelStyle(font, Color.GOLD)
        )
        labelLifeR = Label(
            String.format(Locale.getDefault(), GameConstants.HUD_LABEL_FORMAT_INTEGER, life),
            LabelStyle(font, Color.GOLD)
        )
    }

    /* Creates the table that will hold all labels that will be shown in the heads-up display */
    private fun createLabelsTable(): Table {
        val table = Table()
        table.left().top()
        table.setFillParent(true)
        table.row()
        table.add(labelScoreL).align(Align.right)
        table.add(labelScoreR).align(Align.left)
        table.row()
        table.add(labelLifeL).align(Align.right)
        table.add(labelLifeR).align(Align.left)
        return table
    }

    /* Returns the font size in pixels */
    private fun getFontSize(): Int {
        var size = GameConstants.DEFAULT_FONT_SIZE

        //check for lower resolutions and adjust the size
        val width = Gdx.graphics.width.toFloat()
        val height = Gdx.graphics.height.toFloat()
        if (width < Gdx.graphics.width.toFloat() || width < Gdx.graphics.height.toFloat() || height < Gdx.graphics.width.toFloat() || height < Gdx.graphics.height.toFloat()) {
            size = GameConstants.LOW_RESOLUTION_FONT_SIZE
        }

        return size
    }

    override fun dispose() {
        stage.dispose()
        font.dispose()
    }
}